geode-core/src/distributedTest/java/org/apache/geode/cache/query/dunit/PdxMultiThreadQueryDUnitTest.java

```diff
@@ -0,0 +1,271 @@
+import org.apache.geode.cache.query.TypeMismatchException;
+import org.apache.geode.logging.internal.log4j.api.LogService;
+import org.apache.geode.pdx.PdxReader;
+import org.apache.geode.pdx.PdxSerializable;
+import org.apache.geode.pdx.PdxSerializationException;
+import org.apache.geode.pdx.PdxWriter;
+import org.apache.geode.test.dunit.AsyncInvocation;
+import org.apache.geode.test.dunit.Host;
+import org.apache.geode.test.dunit.NetworkUtils;
+import org.apache.geode.test.dunit.VM;
+import org.apache.geode.test.junit.categories.OQLQueryTest;
+import org.apache.geode.test.version.VersionManager;
+
+@Category({OQLQueryTest.class})
+public class PdxMultiThreadQueryDUnitTest extends PDXQueryTestBase {
+  public static final Logger logger = LogService.getLogger();
+
+  static final int numberOfEntries = 100;
+
+  public PdxMultiThreadQueryDUnitTest() {
+    super();
+  }
+
+  @Test
+  public void testClientServerQuery() throws CacheException {
+    final Host host = Host.getHost(0);
+    final String hostName = NetworkUtils.getServerHostName();
+
+    VM vm0 = host.getVM(VersionManager.CURRENT_VERSION, 0);
+    VM vm1 = host.getVM(VersionManager.CURRENT_VERSION, 1);
+    VM vm2 = host.getVM(VersionManager.CURRENT_VERSION, 2);
+    VM vm3 = host.getVM(VersionManager.CURRENT_VERSION, 3);
+
+    // Start servers
+    for (VM vm : Arrays.asList(vm0, vm1, vm2)) {
+      vm.invoke(() -> {
+        configAndStartBridgeServer();
+      });
+    }
+
+    // create pdx instance at servers
+    vm0.invoke(() -> {
+      Region region = getRootRegion().getSubregion(regionName);
+      for (int i = 0; i < numberOfEntries; i++) {
+        region.put("key-" + i, new TestObject(i, "vmware"));
+      }
+    });
+
+    final int port0 = vm0.invoke(() -> PdxQueryDUnitTest.getCacheServerPort());
+    final int port1 = vm1.invoke(() -> PdxQueryDUnitTest.getCacheServerPort());
+    final int port2 = vm2.invoke(() -> PdxQueryDUnitTest.getCacheServerPort());
+
+    // Create client region
+    vm3.invoke(() -> {
+      ClientCacheFactory cf = new ClientCacheFactory();
+      cf.addPoolServer(hostName, port1);
+      cf.addPoolServer(hostName, port2);
+      cf.addPoolServer(hostName, port0);
+      cf.setPdxReadSerialized(false);
+      ClientCache clientCache = getClientCache(cf);
+      Region region =
+          clientCache.createClientRegionFactory(ClientRegionShortcut.PROXY).create(regionName);
+
+      logger.info("### Executing Query on server: " + queryString[1] + ": from client region: "
+          + region.getFullPath());
+      final int size = 100;
+      int[] array = new int[size];
+      for (int i = 0; i < size; i++) {
+        array[i] = i;
+      }
+      Arrays.stream(array).parallel().forEach(a -> {
+        try {
+          SelectResults rs2 = region.query(queryString[1]);
+          assertEquals(numberOfEntries, rs2.size());
+        } catch (FunctionDomainException e) {
+          e.printStackTrace();
```

Comment:
If you expect to see these exceptions I recommend asserting that expectation.

If you do not expect to see them then I recommend eliminating these catch blocks and allowing the exceptions to cause the test to fail.

Oh but you can't do that because the functional interface doesn't allow exceptions!ions

Would it make sense to have the catch blocks collect these exceptions e.g. into a concurrent list, so the collection could be asserted empty before the test finishes?

Reply:
I added fail("Unexpected query exception:"+e.getMessage());